File: IntelPresentMon/AppWinUI/PresentMonUI.Core/Introspection.cs
namespace PresentMon.UI.Core;

public enum MetricDeviceType { Independent, GraphicsAdapter, System }
public sealed record MetricDeviceAvailability(int DeviceId, int ArraySize, int AvailabilityId);
public sealed record Adapter(int Id, string Vendor, string Name);
public sealed record MetricStat(int Id, string Name, string ShortName, string Description);
public sealed record MetricUnit(int Id, string Name, string ShortName, string Description);
public sealed record MetricAvailabilityReason(int Id, string Description);

public sealed record Metric
{
    public int Id { get; init; }
    public string Name { get; init; } = "";
    public string Description { get; init; } = "";
    public IReadOnlyList<int> AvailableStatIds { get; init; } = [];
    public int PreferredUnitId { get; init; }
    public MetricDeviceType DeviceType { get; init; }
    public IReadOnlyList<MetricDeviceAvailability> DeviceAvailability { get; init; } = [];
    public bool Numeric { get; init; }
}

public sealed record IntrospectionData
{
    public IReadOnlyList<Metric> Metrics { get; init; } = [];
    public IReadOnlyList<MetricStat> Stats { get; init; } = [];
    public IReadOnlyList<MetricUnit> Units { get; init; } = [];
    public IReadOnlyList<Adapter> Adapters { get; init; } = [];
    public int SystemDeviceId { get; init; }
    public int DefaultAdapterId { get; init; }
    public IReadOnlyList<MetricAvailabilityReason> MetricAvailabilityReasons { get; init; } = [];
}

public static class MetricResolver
{
    public const int SystemDeviceId = 65536;
    public static int ArraySize(Metric metric, int deviceId) =>
        metric.DeviceAvailability.FirstOrDefault(d => d.DeviceId == deviceId)?.ArraySize ?? 0;

    public static int ClampArrayIndex(Metric metric, int deviceId, int index) =>
        Math.Clamp(index, 0, Math.Max(0, ArraySize(metric, deviceId) - 1));

    public static int ResolveDeviceId(Metric metric, QualifiedMetric qualified, IntrospectionData intro, Preferences preferences) =>
        metric.DeviceType switch
        {
            MetricDeviceType.Independent => 0,
            MetricDeviceType.System => intro.SystemDeviceId != 0 ? intro.SystemDeviceId : SystemDeviceId,
            MetricDeviceType.GraphicsAdapter => preferences.EnablePerMetricDeviceSelection
                && qualified.DeviceId is not null and not 0 ? qualified.DeviceId.Value : preferences.AdapterId,
            _ => qualified.DeviceId ?? 0,
        };

    public static void Normalize(Metric metric, QualifiedMetric qualified, IntrospectionData intro, Preferences preferences)
    {
        if (metric.DeviceType == MetricDeviceType.Independent) qualified.DeviceId = 0;
        else if (metric.DeviceType == MetricDeviceType.System)
            qualified.DeviceId = intro.SystemDeviceId != 0 ? intro.SystemDeviceId : SystemDeviceId;
        else if (metric.DeviceType == MetricDeviceType.GraphicsAdapter && qualified.DeviceId == 0)
            qualified.DeviceId = null;
        qualified.ArrayIndex = ClampArrayIndex(metric, ResolveDeviceId(metric, qualified, intro, preferences), qualified.ArrayIndex);
        qualified.DesiredUnitId = metric.PreferredUnitId;
    }

    public static QualifiedMetric CreateQualified(Metric metric, IntrospectionData intro, Preferences preferences, int? statId = null)
    {
        var qualified = new QualifiedMetric
        {
            MetricId = metric.Id,
            StatId = statId is not null && metric.AvailableStatIds.Contains(statId.Value)
                ? statId.Value : metric.AvailableStatIds.FirstOrDefault(),
            DeviceId = metric.DeviceType == MetricDeviceType.GraphicsAdapter ? null : 0,
        };
        Normalize(metric, qualified, intro, preferences);
        return qualified;
    }

    public static QualifiedMetric CreateDefault(IntrospectionData intro, Preferences preferences)
    {
        var metric = intro.Metrics.FirstOrDefault(m => m.Id == 8 && m.Numeric)
            ?? intro.Metrics.FirstOrDefault(m => m.Numeric)
            ?? throw new InvalidOperationException("No numeric metrics are available.");
        return CreateQualified(metric, intro, preferences, 1);
    }

    public static bool IsAvailable(Metric metric, QualifiedMetric qualified, IntrospectionData intro, Preferences preferences)
    {
        var deviceId = ResolveDeviceId(metric, qualified, intro, preferences);
        var entry = metric.DeviceAvailability.FirstOrDefault(d => d.DeviceId == deviceId);
        return entry is { AvailabilityId: 0 } && qualified.ArrayIndex >= 0
            && (entry.ArraySize <= 0 ? qualified.ArrayIndex == 0 : qualified.ArrayIndex < entry.ArraySize);
    }

    public static string? AvailabilityReason(Metric metric, QualifiedMetric qualified, IntrospectionData intro, Preferences preferences)
    {
        if (IsAvailable(metric, qualified, intro, preferences)) return null;
        var deviceId = ResolveDeviceId(metric, qualified, intro, preferences);
        var availabilityId = metric.DeviceAvailability.FirstOrDefault(d => d.DeviceId == deviceId)?.AvailabilityId ?? 1;
        return intro.MetricAvailabilityReasons.FirstOrDefault(r => r.Id == availabilityId)?.Description
            ?? "This metric is not available on the selected device.";
    }
}

File: IntelPresentMon/AppWinUICpp/Core/Introspection.cpp
// Copyright (C) 2026 Intel Corporation
// SPDX-License-Identifier: MIT
#include "Introspection.h"

#include <algorithm>
#include <stdexcept>

namespace pmon::ui::core
{
    namespace
    {
        const MetricDeviceAvailability* FindDevice(const Metric& metric, int deviceId)
        {
            const auto it = std::find_if(metric.DeviceAvailability.begin(), metric.DeviceAvailability.end(),
                [deviceId](const auto& candidate) { return candidate.DeviceId == deviceId; });
            return it == metric.DeviceAvailability.end() ? nullptr : &*it;
        }

        bool Contains(const std::vector<int>& values, int value)
        {
            return std::find(values.begin(), values.end(), value) != values.end();
        }
    }

    int MetricResolver::ArraySize(const Metric& metric, int deviceId)
    {
        const auto* entry = FindDevice(metric, deviceId);
        return entry ? entry->ArraySize : 0;
    }

    int MetricResolver::ClampArrayIndex(const Metric& metric, int deviceId, int index)
    {
        return std::clamp(index, 0, std::max(0, ArraySize(metric, deviceId) - 1));
    }

    int MetricResolver::ResolveDeviceId(const Metric& metric, const QualifiedMetric& qualified,
        const IntrospectionData& intro, const Preferences& preferences)
    {
        switch (metric.DeviceType) {
        case MetricDeviceType::Independent:
            return 0;
        case MetricDeviceType::System:
            return intro.SystemDeviceId != 0 ? intro.SystemDeviceId : SystemDeviceId;
        case MetricDeviceType::GraphicsAdapter:
            return preferences.EnablePerMetricDeviceSelection && qualified.DeviceId && *qualified.DeviceId != 0
                ? *qualified.DeviceId : preferences.AdapterId;
        }
        return qualified.DeviceId.value_or(0);
    }

    void MetricResolver::Normalize(const Metric& metric, QualifiedMetric& qualified,
        const IntrospectionData& intro, const Preferences& preferences)
    {
        if (metric.DeviceType == MetricDeviceType::Independent) {
            qualified.DeviceId = 0;
        }
        else if (metric.DeviceType == MetricDeviceType::System) {
            qualified.DeviceId = intro.SystemDeviceId != 0 ? intro.SystemDeviceId : SystemDeviceId;
        }
        else if (metric.DeviceType == MetricDeviceType::GraphicsAdapter && qualified.DeviceId.value_or(0) == 0) {
            qualified.DeviceId.reset();
        }
        qualified.ArrayIndex = ClampArrayIndex(metric, ResolveDeviceId(metric, qualified, intro, preferences), qualified.ArrayIndex);
        qualified.DesiredUnitId = metric.PreferredUnitId;
    }

    QualifiedMetric MetricResolver::CreateQualified(const Metric& metric, const IntrospectionData& intro,
        const Preferences& preferences, std::optional<int> statId)
    {
        QualifiedMetric qualified;
        qualified.MetricId = metric.Id;
        qualified.StatId = statId && Contains(metric.AvailableStatIds, *statId)
            ? *statId : (metric.AvailableStatIds.empty() ? 0 : metric.AvailableStatIds.front());
        qualified.DeviceId = metric.DeviceType == MetricDeviceType::GraphicsAdapter
            ? std::nullopt : std::optional{ 0 };
        Normalize(metric, qualified, intro, preferences);
        return qualified;
    }

    QualifiedMetric MetricResolver::CreateDefault(const IntrospectionData& intro, const Preferences& preferences)
    {
        const auto preferred = std::find_if(intro.Metrics.begin(), intro.Metrics.end(),
            [](const auto& metric) { return metric.Id == 8 && metric.Numeric; });
        const auto numeric = preferred != intro.Metrics.end() ? preferred : std::find_if(intro.Metrics.begin(), intro.Metrics.end(),
            [](const auto& metric) { return metric.Numeric; });
        if (numeric == intro.Metrics.end()) {
            throw std::invalid_argument("No numeric metrics are available.");
        }
        return CreateQualified(*numeric, intro, preferences, 1);
    }

    bool MetricResolver::IsAvailable(const Metric& metric, const QualifiedMetric& qualified,
        const IntrospectionData& intro, const Preferences& preferences)
    {
        const auto* entry = FindDevice(metric, ResolveDeviceId(metric, qualified, intro, preferences));
        return entry && entry->AvailabilityId == 0 && qualified.ArrayIndex >= 0
            && (entry->ArraySize <= 0 ? qualified.ArrayIndex == 0 : qualified.ArrayIndex < entry->ArraySize);
    }

    std::optional<std::string> MetricResolver::AvailabilityReason(const Metric& metric,
        const QualifiedMetric& qualified, const IntrospectionData& intro, const Preferences& preferences)
    {
        if (IsAvailable(metric, qualified, intro, preferences)) {
            return std::nullopt;
        }
        const auto* entry = FindDevice(metric, ResolveDeviceId(metric, qualified, intro, preferences));
        const auto availabilityId = entry ? entry->AvailabilityId : 1;
        const auto reason = std::find_if(intro.MetricAvailabilityReasons.begin(), intro.MetricAvailabilityReasons.end(),
            [availabilityId](const auto& candidate) { return candidate.Id == availabilityId; });
        return reason == intro.MetricAvailabilityReasons.end()
            ? "This metric is not available on the selected device." : reason->Description;
    }
}
